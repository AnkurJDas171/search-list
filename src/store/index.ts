import { configureStore } from "@reduxjs/toolkit";

import usersReducer from "./slice/users.slice";
import listReducer from "./slice/list.slice";
import alertReducer from "./slice/alert.slice";

const store = configureStore({
    reducer: {
        alert: alertReducer,
        users: usersReducer,
        list: listReducer
    }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export default store;
