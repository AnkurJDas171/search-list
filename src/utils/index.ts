import { Dimensions, PixelRatio } from "react-native";
import Fuse from "fuse.js";

import { ListState, UserState, UserType } from "../store/state/type";

const {
    height: HEIGHT,
    width: WIDTH
} = Dimensions.get("window");

console.log(HEIGHT, WIDTH);

const heightBaseScale = HEIGHT / 667;
const widthBaseScale = WIDTH / 375;

export const fontSize = (size: number): number => {
    const newSize = size * widthBaseScale;

    return Math.round(PixelRatio.roundToNearestPixel(newSize));
}

export const height = (height: number): number => {
    const newHeight = height * widthBaseScale;

    return Math.round(PixelRatio.roundToNearestPixel(newHeight));
}

export const fuzzyMatch = (searchedName: string, users: UserState): ListState => {
    const fuse = new Fuse(users, {
        ignoreLocation: false,
        includeScore: true,
        isCaseSensitive: false,
        keys: ["name"],
        threshold: 0.3,
    });

    const searchedResult = fuse.search(searchedName).map(data => data.item);
    return searchedResult
}

const createPairCombination = (str: string): string[] => {
    const pair = [];
    const numberOfPairs = str.split(" ").length;

    for (let i = 0; i < numberOfPairs; i++) {
        let text = "";

        for (let j = i; j < numberOfPairs; j++) {
            text += (text ? " " : "") + str.split(" ")[j];
            pair.push(text);
        }
    }

    return pair;
}

export const isStringPresent = (name: string, searchedName: string): boolean => {
    if (!name) return false;
    if (name.toLocaleLowerCase() === searchedName.toLocaleLowerCase()) return true;

    let isPreset = false;

    const namePairs = createPairCombination(name.toLocaleLowerCase());
    namePairs.forEach((subStr: string) => {
        if (subStr === searchedName.toLocaleLowerCase()) {
            isPreset = true;
            return;
        }
    })

    return isPreset;
}

export const isUserPresent = (searchedName: string, users: UserState): boolean => {
    let userPresent = false;
    if (users.length === 0 || !searchedName) return userPresent;

    users.forEach((user: UserType) => {
        if (user.name && isStringPresent(user.name.toLocaleLowerCase(), searchedName.toLocaleLowerCase())) {
            userPresent = true;
            return;
        }
    })

    return userPresent;
}

export const isUserPresentInFuzzyMatch = (searchedName: string, users: UserState): boolean => {
    if (users.length === 0 || !searchedName) return false;

    const fuzzyMatchData = fuzzyMatch(searchedName, users);
    if (fuzzyMatchData.length > 0) {
        return true;
    }

    return false;
}
