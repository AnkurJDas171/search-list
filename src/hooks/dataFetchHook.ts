import React, { useEffect } from "react";

import { ERROR_ALERT_MESSAGE, dataUrl } from "../assets/Constants";
import { useAppDispatch } from "./reduxHooks";
import { getUsers } from "../store/slice/users.slice";
import { setAlert } from "../store/slice/alert.slice";
import store from "../store";

const fetchUserData = async (storeData: typeof store.dispatch): Promise<void> => {
    try{
        const response = await fetch(dataUrl);
        const data = await response.json();
        storeData(getUsers(data));
    }catch(err){
        console.log("error is: ", err);
        storeData(setAlert(ERROR_ALERT_MESSAGE))
    }
}

const useDataFetch = () => {
    const dispatch = useAppDispatch();

    useEffect(() => {
        fetchUserData(dispatch)
    }, [])
}

export default useDataFetch;
