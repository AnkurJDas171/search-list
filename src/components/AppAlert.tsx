import React, { useEffect } from "react";
import { Animated, Modal, StyleSheet, Text, View } from "react-native";

import colors from "../assets/colors";
import { useAppDispatch, useAppSelector } from "../hooks";
import { RootState } from "../store";
import { AlertType } from "../store/state/type";
import { resetAlert } from "../store/slice/alert.slice";
import { ALERT_VISIBLE_TIME } from "../assets/Constants";

const AppAlert = (): JSX.Element => {
    const dispatch = useAppDispatch();
    const alertState = useAppSelector((state: RootState): AlertType => state.alert)
    const { showAlert, alertText } = alertState;

    const handleRequestClose = () => {
        dispatch(resetAlert());
    }

    useEffect(() => {
        if(showAlert){
            setTimeout(() => {
                handleRequestClose();
            }, ALERT_VISIBLE_TIME)
        }
    }, [showAlert])

    return (
        <Modal
            visible={showAlert}
            animationType="fade"
            onRequestClose={handleRequestClose}
            transparent={true}
        >
            <View style={styles.body}>
                <Animated.View style={[styles.container]}>
                    <Text style={styles.text}>{alertText}</Text>
                </Animated.View>
            </View>
        </Modal>
    )
}

const styles = StyleSheet.create({
    body: {
        height: '100%',
        backgroundColor: colors.BLACK_SEMI_TRANSPARENT,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center'
    },
    container: {
        backgroundColor: colors.WHITE,
        borderRadius: 10,
        height: '15%',
        justifyContent: 'center',
        width: '70%',
        marginBottom: '20%'
    },
    text: {
        fontSize: 15,
        paddingHorizontal: 20,
        textAlign: 'center',
    }
})

export default AppAlert;
