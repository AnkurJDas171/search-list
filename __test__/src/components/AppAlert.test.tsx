import React from "react";
import { render } from "@testing-library/react-native";

import AppAlert from "../../../src/components/AppAlert";
import * as hooks from "../../../src/hooks";

jest.mock("../../../src/hooks", () => ({
    useAppDispatch: jest.fn(),
    useAppSelector: jest.fn(),
}))

describe('Alert component', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    })

    it("should render Alert when showAlert state is true", () => {
        jest.spyOn(hooks, 'useAppSelector').mockReturnValue({
            showAlert: true,
            alertText: 'alert test'
        })

        const screen = render(<AppAlert />)
        const text = screen.getByText('alert test')

        expect(screen.toJSON()).toMatchSnapshot();
        expect(text).toBeTruthy();
    });

    it("should not render Alert is showAlert state is false", () => {
        jest.spyOn(hooks, 'useAppSelector').mockReturnValue({
            showAlert: false,
            alertText: 'alert test'
        });

        const screen: any = render(<AppAlert />)

        expect(screen.toJSON()).toMatchSnapshot();
        expect(screen.toJSON().children).toBeNull()
    });
})
