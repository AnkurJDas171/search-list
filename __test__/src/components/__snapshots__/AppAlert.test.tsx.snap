// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert component should not render Alert when showAlert state is false 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={false}
/>
`;

exports[`Alert component should render Alert when showAlert state is true 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.2)",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 10,
          "height": "15%",
          "justifyContent": "center",
          "marginBottom": "20%",
          "width": "70%",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 15,
            "paddingHorizontal": 20,
            "textAlign": "center",
          }
        }
      >
        alert test
      </Text>
    </View>
  </View>
</Modal>
`;
